src/generate_cropped_dataset.py

```diff
@@ -0,0 +1,335 @@
+"""
+Load a dataset of historic documents by specifying the folder where its located.
+"""
+
+# Utils
+import itertools
+import logging
+import math
+import os
+import os.path
+from datetime import datetime
+from pathlib import Path
+
+import numpy as np
+from PIL import Image
+from torchvision.transforms import functional as F
+from torchvision.utils import save_image
+from tqdm import tqdm
+
+IMG_EXTENSIONS = ['.jpg', '.jpeg', '.png', '.ppm', '.bmp', '.pgm']
+
+
+def has_extension(filename, extensions):
+    """Checks if a file is an allowed extension.
+
+    Adapted from https://github.com/pytorch/vision/blob/master/torchvision/datasets/folder.py.
+
+    Parameters
+    ----------
+    filename : string
+        path to a file
+    extensions : list
+        extensions to match against
+    Returns
+    -------
+    bool
+        True if the filename ends with one of given extensions, false otherwise.
+    """
+    filename_lower = filename.lower()
+    return any(filename_lower.endswith(ext) for ext in extensions)
+
+
+def pil_loader(path, to_rgb=True):
+    pic = Image.open(path)
+    if to_rgb:
+        pic = convert_to_rgb(pic)
+    return pic
+
+
+def convert_to_rgb(pic):
+    if pic.mode == "RGB":
+        pass
+    elif pic.mode in ("CMYK", "RGBA", "P"):
+        pic = pic.convert('RGB')
+    elif pic.mode == "I":
+        img = (np.divide(np.array(pic, np.int32), 2 ** 16 - 1) * 255).astype(np.uint8)
+        pic = Image.fromarray(np.stack((img, img, img), axis=2))
+    elif pic.mode == "I;16":
+        img = (np.divide(np.array(pic, np.int16), 2 ** 8 - 1) * 255).astype(np.uint8)
+        pic = Image.fromarray(np.stack((img, img, img), axis=2))
+    elif pic.mode == "L":
+        img = np.array(pic).astype(np.uint8)
+        pic = Image.fromarray(np.stack((img, img, img), axis=2))
+    else:
+        raise TypeError(f"unsupported image type {pic.mode}")
+    return pic
+
+
+def get_gt_data_paths(directory):
+    """
+    Parameters
+    ----------
+    directory: string
+        parent directory with gt and data folder inside
+
+    Returns
+    -------
+    paths: list of tuples
+
+    """
+
+    paths = []
+    directory = os.path.expanduser(directory)
+
+    path_imgs = os.path.join(directory, "data")
+    path_gts = os.path.join(directory, "gt")
+
+    if not (os.path.isdir(path_imgs) or os.path.isdir(path_gts)):
```

Comment:
Why not using the Patch object as you already use it in other positions

Reply:
legacy implementation